### ISSUE TITLE
Error Hooks for tracer.trace() and tracer.wrap()

### ISSUE THREAD

**Issue by benjohnson** [DataDog/dd-trace-js#2192]
Hello! We've got some custom instrumentation that uses `tracer.wrap()`. One of the challenges we've had is that we want to avoid setting the error tag for client-side errors. To fix this, we've ended up basically forking the code for `tracer.wrap` so that we can control whether or not `tracer.trace` adds the error tags.

Would you be open to a PR that adds some kind of hook or option to control errors?

**Comment by rochdev** [comment]
> Would you be open to a PR that adds some kind of hook or option to control errors?

Definitely, we're always open to contributions! However, can you also provide some code snippet that demonstrates better your use case? It's possible there is an existing way to handle this. For example we handle this in our own auto-instrumentation for HTTP servers by ignoring 4xx (and thus client) errors.

**Comment by benjohnson** [comment]
In your HTTP server config, you control how you finish the span, which allows you to call [addStatusError](https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/plugins/util/web.js#L260) and do the check.

If you're using the out of the box `tracer.wrap()` the finishing of the span happens [inside tracer.trace()](https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/plugins/util/web.js#L260) with the call to [addError](https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/tracer.js#L139)

We could monkeypatch the tracer to do this but, a) it's private, and b) we only want to apply this logic on some of our spans.

---

Here's our use case:

We want an additional named span to occur for every function call in a file. This means we're calling code like this:

```js
      for (const [fnName, fn] of Object.entries(service)) {
        if (typeof fn === 'function') {
          service[fnName] = wrap(
            'cookiemonster',
            { resource: `service.cookiemonster.${fnName}`, tags },
            fn
          );
        }
      }
```

With this basic use of wrap, the wrap instrumentation is responsible for finishing the spans, so by the time we see the result of this call, we've already marked any lower level errors as having the error tag. There's no way for us to "undo" it.

I can't think of a way for us to modify this without being able to control the tracer's `addError` method.

In our internal code, we've basically copied what `wrap` and `trace` are doing and are using the lower level functions, but we're seeing some weird behaviour in DD. For instance, when we turned this on our downstream service tracing dropped to 0:

<img width="600" alt="image" src="https://user-images.githubusercontent.com/69549/179756974-f6880fb1-8163-48c0-ad9a-74398896f070.png">

I understand the desire to not overcomplicate settings, but I also would really like a way to control which things we mark as errors so that we can continue to benefit from your changes and updates to `wrap`.

Thanks for considering.

**Comment by benjohnson** [comment]
Yep, I do think that this would work for us. Is it better to wait until you've had a chance to make the changes you want to make to the tracer before putting together a PR?

**Comment by rochdev** [comment]
@benjohnson It really depends how this is implemented, but if it would be with a hook I think it would be fine either way since only the body of the hook would change, not the hook itself. However, it might be better to do this through `diagnostics_channel` which we're moving a lot of things to right now, and if that's used it might be better to wait for the new code to land since it will standardize the naming conventions for the channel events.

**Comment by yarinvak** [comment]
bumping it, we will also appreciate this feature
